### ISSUE TITLE
MaxListenersExceededWarning on wrap()

### ISSUE THREAD

**Issue by yukha-dw** [connor4312/cockatiel#69]
Hello, I have question on how to properly use `wrap()`.

On [this guide](https://github.com/connor4312/cockatiel#wrappolicies),
```
const result = await retry.execute(() =>
  breaker.execute(() => timeout.execute(({ signal }) => getData(signal))),
);
```
Is the equivalent to:
```
const result = await wrap(retry, breaker, timeout).execute(({ signal }) => getData(signal));
```

But, when I test it, I got `MaxListenersExceededWarning` when using `wrap()`.

My code:
```
const { wrap, handleAll, retry, circuitBreaker, timeout, TimeoutStrategy, ExponentialBackoff, ConsecutiveBreaker } = require('cockatiel');

const getData = async (signal) => {
    return { signal }
}

const retry1 = retry(handleAll, { maxAttempts: 3, backoff: new ExponentialBackoff() });
const circuitBreaker1 = circuitBreaker(handleAll, {
    halfOpenAfter: 10 * 1000,
    breaker: new ConsecutiveBreaker(5),
});
const timeout1 = timeout(2000, TimeoutStrategy.Cooperative);

const safe = async () => {
    let result;
    console.log("SAFE START");
    for (let i = 0; i < 11; i++) {
        result = await retry1.execute(() =>
            circuitBreaker1.execute(() => timeout1.execute(({ signal }) => getData(signal))),
        );
    }
    await new Promise(r => setTimeout(r, 1000));
    console.log("SAFE END");
}

const unsafe = async () => {
    let result;
    console.log("UNSAFE START");
    for (let i = 0; i < 11; i++) {
        result = await wrap(retry1, circuitBreaker1, timeout1).execute(({ signal }) => getData(signal));
    }
    await new Promise(r => setTimeout(r, 1000));
    console.log("UNSAFE END");
}

const main = async () => {
    await safe();
    await new Promise(r => setTimeout(r, 1000));
    await unsafe();
}

main();
```

Output:
```
SAFE START
SAFE END
UNSAFE START
(node:9672) MaxListenersExceededWarning: Possible EventTarget memory leak detected. 11 abort listeners added to [AbortSignal]. Use events.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
UNSAFE END
```

Is it a leak or working as intended? Thank you.

**Comment by clement-heliot** [comment]
We are experimenting with Cockatiel on one of our high-traffic services and observing the same behavior. Seems like using `wrap` introduce a memory leak.

**Comment by connor4312** [comment]
Thanks for the ping, will fix this today

**Comment by JonasFaure** [comment]
Hi @connor4312 

From what I understood the issue comes from the usage of never unreferenced [_neverAbortedSignal_](https://github.com/connor4312/cockatiel/blob/e4a8111dda017f95573710cdc1834d29dac638cc/src/common/abort.ts#L3) as the default value for the signal in [RetryPolicy](https://github.com/connor4312/cockatiel/blob/e4a8111dda017f95573710cdc1834d29dac638cc/src/RetryPolicy.ts#L97) & [CircuitBreakerPolicy](https://github.com/connor4312/cockatiel/blob/e4a8111dda017f95573710cdc1834d29dac638cc/src/CircuitBreakerPolicy.ts#L144)

Since wrap() function processes policies passed as parameters from left to right and pass the AbortSignal from one policy to the next we end up with the [TimeoutPolicy](https://github.com/connor4312/cockatiel/blob/e4a8111dda017f95573710cdc1834d29dac638cc/src/TimeoutPolicy.ts#L90) calling `signal.addEventListener` for each execution, leading to an EventEmitter with very large numbers of listeners being kept forever in memory.

<img width="1177" alt="Screenshot 2023-03-14 at 18 11 08" src="https://user-images.githubusercontent.com/47354296/225272416-7f9b3cad-103c-491c-a706-5c7631654c0a.png">
👆 screenshot of a >250Mb AbortSignal object in heapdump

**Comment by connor4312** [comment]
Indeed. Fix published in 3.1.1

**Comment by clement-heliot** [comment]
@connor4312 Thank you for the (very) quick fix!
